### PR TITLE
Large Model: Support offset

### PR DESCRIPTION
**Pull request by snnn**
To support saving multiple tensor data into a single external file with offsets

### CODE REVIEW

File: onnxruntime/core/platform/posix/env.cc [hunk 1]
```diff
@@ -116,31 +116,45 @@ class PosixEnv : public Env {
...
+    if(len > 0) return true;
+    struct stat stbuf;
+    if ((fstat(fd, &stbuf) != 0) || (!S_ISREG(stbuf.st_mode))) {
+      return false;
+    }
+    len = static_cast<size_t>(stbuf.st_size);
+    return true;
+  }
+
+  static bool CanDoMemoryMapping(off_t offset) {
```
Comment (RyanUnderhill) [inline onnxruntime/core/platform/posix/env.cc]: You only do memory mapping if the offset is aligned to a system page boundary? Why not just map and offset the pointer you return to align properly?

File: onnxruntime/core/framework/tensor_external_data_info.h [hunk 2]
```diff
@@ -10,16 +10,23 @@
 namespace onnxruntime {
 class ExternalDataInfo {
  private:
+#ifdef _WIN32
+  using OFFSET_TYPE = int64_t;
```
Comment (RyanUnderhill) [inline onnxruntime/core/framework/tensor_external_data_info.h]: Signed int? Is there any issue if you use uint64_t for all platforms?

File: onnxruntime/core/platform/posix/env.cc [hunk 3]
```diff
@@ -116,31 +116,45 @@ class PosixEnv : public Env {
...
     } while (length_remain > 0);
     p = buffer.release();
-    len = stbuf.st_size;
     deleter.f = DeleteBuffer;
     deleter.param = p;
     return Status::OK();
   }
 
-  common::Status ReadFileAsString(const char* fname, void*& p, size_t& len, OrtCallback& deleter) const override {
+  static bool CalcFileSize(int fd, size_t& len){
```
Comment (RyanUnderhill) [inline onnxruntime/core/platform/posix/env.cc]: Random nitpick, but should it just be called GetFileSize? It doesn't really calculate anything, it just gets the value.

File: onnxruntime/core/platform/windows/env.cc [hunk 4]
```diff
@@ -94,10 +94,30 @@ class WindowsEnv : public Env {
   void ExecuteTask(const Task& t) const override {
     t.f();
   }
-  common::Status ReadFileAsString(const wchar_t* fname, void*& p, size_t& len, OrtCallback& deleter) const override {
+
+  static common::Status CalcFileSize(const wchar_t* fname, HANDLE hFile, size_t& l) {
```
Comment (RyanUnderhill) [inline onnxruntime/core/platform/windows/env.cc]: Should use a better variable name than 'l'. I thought the if below was if(1>0).

File: onnxruntime/core/framework/tensor_external_data_info.h (hunk 2 above)
Comment (snnn) [reply onnxruntime/core/framework/tensor_external_data_info.h]: On Linux, off_t is a signed integer. And, sometimes it's 32 bits, sometimes it's 64 bits. It is very complicated. I'm reading this article: https://android.googlesource.com/platform/bionic/+/master/docs/32-bit-abi.md

File: onnxruntime/core/platform/posix/env.cc [hunk 5]
```diff
@@ -123,7 +128,9 @@ class PosixEnv : public Env {
     auto length_remain = len;
     do {
       size_t bytes_to_read = length_remain;
-      ssize_t bytes_readed = offset > 0 ? pread(fd, wptr, bytes_to_read, offset) : read(fd, wptr, bytes_to_read);
+      ssize_t bytes_readed;
```
Comment (RyanUnderhill) [inline onnxruntime/core/platform/posix/env.cc]: bytes_readed should be just 'bytes_read' (read is the past tense of read, but it's pronounced differently...)

**Pull request merged**